pallets/connectors-gateway/axelar-gateway-precompile/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+	pub type SourceConversion<T: Config> = StorageMap<_, Twox64Concat, H256, SourceConverter>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T> {
+		pub gateway: H160,
+		_phantom: core::marker::PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			GenesisConfig {
+				gateway: Default::default(),
+				_phantom: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			AxelarGatewayContract::<T>::set(self.gateway)
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		GatewaySet {
+			address: H160,
+		},
+		ConverterSet {
+			id_hash: H256,
+			converter: SourceConverter,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The given domain is not yet allowlisted, as we have no converter yet
+		NoConverterForSource,
+		/// A given domain expects a given structure for account bytes and it
+		/// was not given here.
+		AccountBytesMismatchForDomain,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		#[pallet::call_index(0)]
+		pub fn set_gateway(origin: OriginFor<T>, address: H160) -> DispatchResult {
+			<T as Config>::AdminOrigin::ensure_origin(origin)?;
+
+			AxelarGatewayContract::<T>::set(address);
+
+			Self::deposit_event(Event::<T>::GatewaySet { address });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		#[pallet::call_index(1)]
+		pub fn set_converter(
+			origin: OriginFor<T>,
+			id_hash: H256,
+			converter: SourceConverter,
+		) -> DispatchResult {
+			<T as Config>::AdminOrigin::ensure_origin(origin)?;
+
+			SourceConversion::<T>::insert(id_hash, converter.clone());
+
+			Self::deposit_event(Event::<T>::ConverterSet { id_hash, converter });
+
+			Ok(())
+		}
+	}
+}
+
+#[precompile_utils::precompile]
+impl<T: Config> Pallet<T>
+where
+	T: frame_system::Config,
+	<T as frame_system::Config>::RuntimeOrigin: From<pallet_connectors_gateway::GatewayOrigin>,
+{
+	// Mimics:
+	//
+	//   function execute(
+	//         bytes32 commandId,
+	//         string calldata sourceChain,
+	//         string calldata sourceAddress,
+	//         bytes calldata payload
+	//     ) external {
+	//       bytes32 payloadHash = keccak256(payload);
+	// 		 if (
+	//           !gateway.validateContractCall(
+	//              commandId,
+	//              sourceChain,
+	//              sourceAddress,
+	//              payloadHash)
+	//           ) revert NotApprovedByGateway();
+	//
+	//        _execute(sourceChain, sourceAddress, payload);
+	// }
+	//
+	// Note: The _execute logic in this case will forward all calls to the
+	//       pallet-connectors-gateway with a special runtime local origin
+	#[precompile::public("execute(bytes32,string,string,bytes)")]
+	fn execute(
+		handle: &mut impl PrecompileHandle,
+		command_id: H256,
+		source_chain: String<MAX_SOURCE_CHAIN_BYTES>,
```

Comment:
Since all we care about is that these are buckets of bytes that we can toss into the hasher, and don't rely on them being printable UTF-8 strings, I don't think we necessarily need a different `String` type from `Bytes` here. That would also avoid the possible failure on a non-utf-8 source (but would that ever really happen?).

On the other hand, it's nice to match the solidity signature. :shrug:

Simplifying the hashing code by treating these as just bytes might be a touch cleaner, but I think this is fine

Reply:
I will stick with compliant parameters here.